### ISSUE TITLE
Procedure for signing certificates without touching the file system

### ISSUE THREAD

**Issue by mholt** [smallstep/certificates#185]
### What would you like to be added

The ability to issue certificates with [an Authority](https://pkg.go.dev/github.com/smallstep/certificates@v0.14.0-rc.3/authority?tab=doc#New) but without touching the file system, if possible.

### Why this is needed

As you know, I'm working on integrating Smallstep into Caddy. Caddy has [its own storage facilities](https://caddyserver.com/docs/json/storage/). The default `Storage` interface implementation is indeed the file system, but it could be anything -- a database, a network store, or anything else with atomic operations. Even if it is the file system, users would not expect users to write outside of the configured storage path (e.g. a `~/.step` write would likely fail on tightly configured systems).

Basically, a library dependency of Caddy shouldn't require writing to the file system.

I'm trying to use `authority.New` but the [`authority.Config`](https://pkg.go.dev/github.com/smallstep/certificates@v0.14.0-rc.3/authority?tab=doc#Config) type works with file paths. This is problematic, since the resources may not be coming from files.

Can we eliminate the dependency on the file system? Instead, if we could set the decoded root certificate and key directly, for example, a *x509.Certificate field -- that would be excellent. What do you think?

**Comment by maraino** [comment]
Support for this will be available once https://github.com/smallstep/certificates/tree/kms branch is merged

**Comment by mholt** [comment]
Thanks @maraino! I am trying it now, on that branch.

I have a root and intermediate cert (and their keys), and am ready to create an Authority to sign stuff. I'm not quite sure yet about crypto.Signer, but this is what I have so far:


```go
auth, err := authority.New(new(authority.Config),
	authority.WithX509Signer(interCert, interKey.(crypto.Signer)),
)
if err != nil {
	log.Fatal(err)
}
```

But it fails with `address cannot be empty`, from here: https://github.com/smallstep/certificates/blob/43bd8113aaba57b2e64238e9478f34dbc6f5cb71/authority/config.go#L143-L147

For now, my goal is very simple: just get a signed cert from an embedded CA (no file system, no ACME ops, no ports, etc). I figure once I make an `Authority` I can just call `Sign()`.

Any tips? :grin:

**Comment by maraino** [comment]
First, the `interKey.(crypto.Signer)` it's fine, I'm using `crypto.Signer` to enforce an interface that the private keys supports, but if the method that you're using returns `interface{}` instead of the actual type of the key or `crypto.Signer` will require the "cast".

Second, the address won't be used by the authority part, so any placeholder value must be good, or you can also use the address that caddy is using (e.g. `:443`). In the same method, there are other values that cannot be empty, any placeholder (e.g. `"manually configured"`) on them should work, as long as you are using the `Option` method that avoids using the specific parameter.

The only thing that it will be required will be probably at least one provisioner.

**Comment by mholt** [comment]
Alrighty -- I can work with that for now.

I now have:

```go
cfg := &authority.Config{
	Address:          "placeholder_Address:1",
	Root:             []string{"placeholder_Root"},
	IntermediateCert: "placeholder_IntermediateCert",
	IntermediateKey:  "placeholder_IntermediateKey",
	DNSNames:         []string{"placeholder_DNSNames"},
	AuthorityConfig: &authority.AuthConfig{},
}
auth, err := authority.New(cfg,
	authority.WithX509Signer(interCert, interKey.(crypto.Signer)),
)
if err != nil {
	log.Fatal(err)
}
```

But the error is:

```
open placeholder_Root failed: no such file or directory
```

So it is still trying to access the file system. (No difference if I add a Provisioner.) Any way around that?

**Comment by maraino** [comment]
This is because you need to pass a root certificate/s using 
```go
WithX509RootBundle(pemCerts []byte) Option
```

**Comment by maraino** [comment]
I can add something like `WithX509RootCertificate(cert *x509.Certificate) Option` if you prefer

**Comment by mholt** [comment]
Oh. Yeah, I have `*x509.Certificate` values returned from `profile.CreateCertificate()` and then calling `x509.ParseCertificate()` on them. I could PEM-encode them, but I imagine you need it decoded anyway? If so, then I think that'd be a good thing to add.

Why does the Authority need a root certificate when it has a signed intermediate for its signer? (Just curious, trying to get familiar with how Smallstep works.)

**Comment by maraino** [comment]
Added here https://github.com/smallstep/certificates/commit/2d4f369db2f7c45ce580d59167d7f692a9297842

**Comment by mholt** [comment]
Great, thanks @maraino! I was able to get a signed certificate using this spike code:

```go
package main

import (
	"crypto"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/x509"
	"encoding/pem"
	"log"
	"os"

	"github.com/smallstep/certificates/authority"
	"github.com/smallstep/certificates/authority/provisioner"
	"github.com/smallstep/cli/crypto/x509util"
)

func main() {
	caName := "Testing CA"

	rootCert, rootKey, err := newRoot(caName)
	if err != nil {
		log.Fatal(err)
	}

	interCert, interKey, err := newIntermediate(caName, rootCert, rootKey)
	if err != nil {
		log.Fatal(err)
	}

	cfg := &authority.Config{
		Address:          "placeholder_Address:1",
		Root:             []string{"placeholder_Root"},
		IntermediateCert: "placeholder_IntermediateCert",
		IntermediateKey:  "placeholder_IntermediateKey",
		DNSNames:         []string{"placeholder_DNSNames"},
		AuthorityConfig: &authority.AuthConfig{},
		},
	}
	auth, err := authority.New(cfg,
		authority.WithX509Signer(interCert, interKey.(crypto.Signer)),
		authority.WithX509RootCerts(rootCert),
	)
	if err != nil {
		log.Fatal(err)
	}

	priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		log.Fatal(err)
	}

	template := &x509.CertificateRequest{
		DNSNames: []string{"dns.name"},
	}

	csr, err := x509.CreateCertificateRequest(rand.Reader, template, priv)
	if err != nil {
		log.Fatal(err)
	}

	cr, err := x509.ParseCertificateRequest(csr)
	if err != nil {
		log.Fatal(err)
	}

	cert, err := auth.Sign(cr, provisioner.Options{})
	if err != nil {
		log.Fatal(err)
	}

	f, err := os.Create("/Users/matt/Desktop/leaf_cert.pem")
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	for i, c := range cert {
		pem.Encode(f, &pem.Block{
			Type:  "CERTIFICATE",
			Bytes: cert[0].Raw,
		})
	}
}

func newRoot(caName string) (rootCrt *x509.Certificate, privateKey interface{}, err error) {
	rootProfile, err := x509util.NewRootProfile(caName)
	if err != nil {
		return
	}
	return newCert(rootProfile)
}

func newIntermediate(caName string, rootCrt *x509.Certificate, rootKey interface{}) (cert *x509.Certificate, privateKey interface{}, err error) {
	interProfile, err := x509util.NewIntermediateProfile(caName, rootCrt, rootKey)
	if err != nil {
		return
	}
	return newCert(interProfile)
}

func newCert(profile x509util.Profile) (cert *x509.Certificate, privateKey interface{}, err error) {
	certBytes, err := profile.CreateCertificate()
	if err != nil {
		return
	}
	privateKey = profile.SubjectPrivateKey()
	cert, err = x509.ParseCertificate(certBytes)
	return
}

```

Haven't tested this with an actual TLS client yet, but that's the next phase of my experiments.

The placeholder values are a little unfortunate -- if you do find an opportunity to not require them, it would be nice/cleaner -- but I think this will work for now. Obviously you know the code base way better; if you notice anything I should be doing differently, just let me know! (I'm just starting with the basics and will build it up and out from here.)

**Comment by mholt** [comment]
Actually, this code still creates a folder: `~/.step` (I think when a profile is created). Do I need to use some [WithOptions](https://pkg.go.dev/github.com/smallstep/cli/crypto/x509util?tab=doc#WithOption)?

**Comment by maraino** [comment]
Great you got it working. The `~/.step` directory is automatically created by an init function at `github.com/smallstep/cli/config`, this is probably required by one of our packages. Right now we don't have a way to disable it.

**Comment by mholt** [comment]
Okie. That's fine for now; not a showstopper at this point. I'll leave this issue open until there's a way to turn that off, since we'll need that before an initial mainline release in Caddy.

Thanks for your help!

**Comment by maraino** [comment]
The issue described is already fixed. 
The issue with the STEPPATH will be tracked with https://github.com/smallstep/cli/issues/223
